<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>The `GeDi` User's Guide
 • GeDi</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="The `GeDi` User's Guide
">
<meta property="og:description" content="GeDi">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GeDi</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/GeDi_manual.html">The `GeDi` User's Guide
</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/AnnekathrinSilvia/GeDi" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="GeDi_manual_files/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="GeDi_manual_files/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip></h1>
<p>The <code>GeDi</code> User’s Guide</p>
                        <h4 data-toc-skip class="author">Annekathrin
Silvia Nedwed</h4>
            <address class="author_afil">
      Institute of Medical Biostatistics, Epidemiology and Informatics
(IMBEI), Mainz<br><a class="author_email" href="mailto:#"></a><a href="mailto:anneludt@uni-mainz.de" class="email">anneludt@uni-mainz.de</a>
      </address>
                              <h4 data-toc-skip class="author">Federico
Marini</h4>
            <address class="author_afil">
      Institute of Medical Biostatistics, Epidemiology and Informatics
(IMBEI), MainzCenter for Thrombosis and Hemostasis (CTH),
Mainz<br><a class="author_email" href="mailto:#"></a><a href="mailto:marinif@uni-mainz.de" class="email">marinif@uni-mainz.de</a>
      </address>
                  
            <h4 data-toc-skip class="date">21 March 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/AnnekathrinSilvia/GeDi/blob/HEAD/vignettes/GeDi_manual.Rmd" class="external-link"><code>vignettes/GeDi_manual.Rmd</code></a></small>
      <div class="hidden name"><code>GeDi_manual.Rmd</code></div>

    </div>

    
    
<style type="text/css">
.smaller {
  font-size: 10px
}
</style>
<p><strong>Compiled date</strong>: 2024-03-21</p>
<p><strong>Last edited</strong>: 2024-02-29</p>
<p><strong>License</strong>:</p>
<hr>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette introduces the usage of the <code>GeDi</code> package
for exploring the results of functional annotation and enrichment
analyses.</p>
<p><code>GeDi</code> is a versatile package designed to simplify the
exploration and comprehension of functional annotation and enrichment
analysis results. It offers a <em><a href="https://CRAN.R-project.org/package=shiny" class="external-link">shiny</a></em>
application that combines interactivity, visualization, and
reproducibility to consolidate comprehensive outcomes.</p>
<p>To incorporate <code>GeDi</code> into your workflow, you’ll need the
results of a functional annotation or enrichment analysis. This vignette
demonstrates the core functionalities of <code>GeDi</code> using a
publicly available dataset from Alasoo et al., as described in their
paper “Shared genetic effects on chromatin and gene expression indicate
a role for enhancer priming in immune response” <span class="citation">(Alasoo et al. 2018)</span>.</p>
<p>Accessible through the <em><a href="https://bioconductor.org/packages/3.19/macrophage" class="external-link">macrophage</a></em>
Bioconductor package, this dataset comprises files generated from Salmon
quantification (version 0.12.0, with Gencode v29 reference) and
gene-level summarized values.</p>
<p>Within the <em><a href="https://bioconductor.org/packages/3.19/macrophage" class="external-link">macrophage</a></em>
experimental setup, samples derive from six different donors under four
distinct conditions: naive, treated with Interferon gamma, with SL1344,
or with a combination of Interferon gamma and SL1344. For illustration,
we will focus on comparing Interferon gamma-treated samples with naive
samples.</p>
</div>
<div class="section level2">
<h2 id="gettingstarted">Getting started<a class="anchor" aria-label="anchor" href="#gettingstarted"></a>
</h2>
<p>Before you can start using GeDi, the package needs to be installed on
your machine. To install the package, begin by opening R and executing
the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"GeDi"</span><span class="op">)</span></span></code></pre></div>
<p>Once installed, the package can be loaded and attached to your
current workspace as follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/AnnekathrinSilvia/GeDi" class="external-link">"GeDi"</a></span><span class="op">)</span></span></code></pre></div>
<p>With the attached package, you can simply start the application by
running <code><a href="../reference/GeDi.html">GeDi()</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GeDi.html">GeDi</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>This action will open the application, directing you to the
<strong>Welcome</strong> page. From there, you can easily provide your
data using the <strong>Data Input</strong> panel on the left side menu,
ensuring it’s in the correct format for analysis.</p>
<p>Alternatively, you can initiate the application by executing:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">GeDi</span>(</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>  <span class="at">genesets =</span> geneset_df,</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="at">ppi =</span> ppi_df</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="at">distance_scores =</span> distance_scores_df</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>)</span></code></pre></div>
<p>where</p>
<ul>
<li>
<code>geneset_df</code> represents your input data in the form of a
<code>DataFrame</code>, which should include at least one column named
“Genesets” containing geneset identifiers and one column named “Genes”
containing a comma-separated list of genes belonging to each respective
geneset.</li>
<li>
<code>ppi_df</code> is another <code>DataFrame</code> containing
protein-protein interaction scores, with columns named “from”, “to”, and
“combined_score”.</li>
<li>
<code>distance_scores_df</code> is a sparse <code>Matrix</code>
containing the distance scores of the genesets in your data.</li>
</ul>
<p>All of these parameters are optional, as you can alternatively
upload, download, and compute them directly within the application.
However, some of these processes may require a significant amount of
time, especially with larger datasets. Therefore, it may be advantageous
to save the intermediate results, such as the downloaded PPI and
computed distance scores, for later use within the application.</p>
<p>In this vignette, we demonstrate the functionality of
<code>GeDi</code> using enrichment analysis results from the <em><a href="https://bioconductor.org/packages/3.19/macrophage" class="external-link">macrophage</a></em>
dataset. To immediately start exploring the application, you can simply
execute:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GeDi.html">GeDi</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>and load the example data with the <code>Load example data</code>
button in the <strong>Data Input</strong> panel.</p>
<p>Alternatively, you can proceed by following the subsequent code
chunks to create the necessary input objects, step by step. This can
serve as a reference guide for the steps ideally executed prior to
analyzing the data with <code>GeDi</code>.</p>
<p>To utilize <code>GeDi</code>, you’ll require results from a
functional annotation analysis. In this vignette, we’ll demonstrate how
to conduct an enrichment analysis on differentially expressed (DE) genes
from the <em><a href="https://bioconductor.org/packages/3.19/macrophage" class="external-link">macrophage</a></em>
dataset.</p>
<p>Firstly, we’ll load the macrophage data and create a
<code>DESeqDataset</code>, as the subsequent differential expression
analysis will be performed using <em><a href="https://bioconductor.org/packages/3.19/DESeq2" class="external-link">DESeq2</a></em>
<span class="citation">(Love, Huber, and Anders 2014)</span>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load required libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"macrophage"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/thelovelab/DESeq2" class="external-link">"DESeq2"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load the example dataset "gse" from the "macrophage" package</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"gse"</span>, package <span class="op">=</span> <span class="st">"macrophage"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a DESeqDataSet object using the "gse" dataset and define the </span></span>
<span><span class="co"># experimental design.</span></span>
<span><span class="co"># We use the condition as part of the experimental design, because we are </span></span>
<span><span class="co"># interested in the differentially expressed genes between treatments. We also </span></span>
<span><span class="co"># add the line to the design to account for the inherent differences between </span></span>
<span><span class="co"># the donors.</span></span>
<span><span class="va">dds_macrophage</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSet</a></span><span class="op">(</span><span class="va">gse</span>, design <span class="op">=</span> <span class="op">~</span> <span class="va">line</span> <span class="op">+</span> <span class="va">condition</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Change the row names of the DESeqDataSet object to Ensembl IDs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dds_macrophage</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\..*"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dds_macrophage</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Have a look at the resulting DESeqDataSet object</span></span>
<span><span class="va">dds_macrophage</span></span>
<span><span class="co">#&gt; class: DESeqDataSet </span></span>
<span><span class="co">#&gt; dim: 58294 24 </span></span>
<span><span class="co">#&gt; metadata(7): tximetaInfo quantInfo ... txdbInfo version</span></span>
<span><span class="co">#&gt; assays(3): counts abundance avgTxLength</span></span>
<span><span class="co">#&gt; rownames(58294): ENSG00000000003 ENSG00000000005 ... ENSG00000285993 ENSG00000285994</span></span>
<span><span class="co">#&gt; rowData names(2): gene_id SYMBOL</span></span>
<span><span class="co">#&gt; colnames(24): SAMEA103885102 SAMEA103885347 ... SAMEA103885308 SAMEA103884949</span></span>
<span><span class="co">#&gt; colData names(15): names sample_id ... condition line</span></span></code></pre></div>
<p>Now that we’ve obtained our <code>DESeqDataset</code>, we can conduct
the differential expression (DE) analysis. In this vignette, we’ll
utilize the results from comparing two distinct conditions of the
dataset, specifically <code>IFNg</code> and <code>naive</code>, while
accounting for the cell line of origin.</p>
<p>Before executing the DE analysis, we’ll filter out lowly expressed
features from the dataset. In this instance, we’ll exclude all genes
with fewer than 10 counts in at least 6 samples, where 6 corresponds to
the smallest group size in the dataset.</p>
<p>Subsequently, we’ll conduct the DE analysis and assess against a null
hypothesis of a log2FoldChange of 1 to ensure that we identify genes
with consistent and robust changes in expression.</p>
<p>Finally, we’ll append the gene symbols to the resultant
<code>DataFrame</code>, which will later serve as our “Genes” column in
the input data for <code>GeDi</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Filter genes based on read counts</span></span>
<span><span class="co"># Calculate the number of genes with at least 10 counts in at least 6 samples</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">dds_macrophage</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">6</span></span>
<span></span>
<span><span class="co"># Subset the DESeqDataSet object to keep only the selected genes</span></span>
<span><span class="va">dds_macrophage</span> <span class="op">&lt;-</span> <span class="va">dds_macrophage</span><span class="op">[</span><span class="va">keep</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Have a look at the resulting DESeqDataSet object</span></span>
<span><span class="va">dds_macrophage</span></span>
<span><span class="co">#&gt; class: DESeqDataSet </span></span>
<span><span class="co">#&gt; dim: 17806 24 </span></span>
<span><span class="co">#&gt; metadata(7): tximetaInfo quantInfo ... txdbInfo version</span></span>
<span><span class="co">#&gt; assays(3): counts abundance avgTxLength</span></span>
<span><span class="co">#&gt; rownames(17806): ENSG00000000003 ENSG00000000419 ... ENSG00000285982 ENSG00000285994</span></span>
<span><span class="co">#&gt; rowData names(2): gene_id SYMBOL</span></span>
<span><span class="co">#&gt; colnames(24): SAMEA103885102 SAMEA103885347 ... SAMEA103885308 SAMEA103884949</span></span>
<span><span class="co">#&gt; colData names(15): names sample_id ... condition line</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform differential expression analysis using DESeq2</span></span>
<span><span class="va">dds_macrophage</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds_macrophage</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract differentially expressed genes</span></span>
<span><span class="co"># Perform contrast analysis comparing "IFNg" condition to "naive" condition</span></span>
<span><span class="co"># Set a log2 fold change threshold of 1 and a significance level (alpha) of 0.05</span></span>
<span><span class="va">res_macrophage_IFNg_vs_naive</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">dds_macrophage</span>,</span>
<span>  contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"IFNg"</span>, <span class="st">"naive"</span><span class="op">)</span>,</span>
<span>  lfcThreshold <span class="op">=</span> <span class="fl">1</span>, alpha <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add gene symbols to the results in a column "SYMBOL"</span></span>
<span><span class="va">res_macrophage_IFNg_vs_naive</span><span class="op">$</span><span class="va">SYMBOL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">dds_macrophage</span><span class="op">)</span><span class="op">$</span><span class="va">SYMBOL</span></span></code></pre></div>
<p>After completing the differential expression analysis, we move on to
conduct the functional annotation analysis. To begin, we extract the
differentially expressed (DE) genes from the previously generated
results and identify the background genes to be utilized for functional
enrichment.</p>
<p>For the enrichment analysis, we use the overrepresentation analysis
method provided by the <em><a href="https://bioconductor.org/packages/3.19/topGO" class="external-link">topGO</a></em>
package. To streamline the integration of these results into
<code>GeDi</code>, we utilize the <code>topGOtable</code> function from
the <em><a href="https://bioconductor.org/packages/3.19/pcaExplorer" class="external-link">pcaExplorer</a></em>
package. By default, this function employs the <code>BP</code> ontology
and the <code>elim</code> method, which helps decorrelate the Gene
Ontology (GO) graph structure, resulting in less redundant functional
categories. The output is a <code>DataFrame</code> object that
seamlessly integrates with <code>GeDi</code>.</p>
<p>However, as <code>GeDi</code> has only minimal requirements for the
input, enrichment results generated using <em><a href="https://bioconductor.org/packages/3.19/clusterProfiler" class="external-link">clusterProfiler</a></em>
can also be utilized. While we primarily tested results from the
<code>enrichGO</code> method during <code>GeDi</code> development, those
from the <code>enrichKEGG</code> and <code>enrichPathway</code> methods
are also compatible.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load required packages for analysis</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/federicomarini/pcaExplorer" class="external-link">"pcaExplorer"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/federicomarini/GeneTonic" class="external-link">"GeneTonic"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/AnnotationDbi" class="external-link">"AnnotationDbi"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract gene symbols from the DESeq2 results object where FDR is below 0.05</span></span>
<span><span class="co"># The function deseqresult2df is used to convert the DESeq2 results to a </span></span>
<span><span class="co"># dataframe format</span></span>
<span><span class="co"># FDR is set to 0.05 to filter significant results</span></span>
<span><span class="va">de_symbols_IFNg_vs_naive</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GeneTonic/man/deseqresult2df.html" class="external-link">deseqresult2df</a></span><span class="op">(</span><span class="va">res_macrophage_IFNg_vs_naive</span>, FDR <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">$</span><span class="va">SYMBOL</span></span>
<span></span>
<span><span class="co"># Extract gene symbols for background using the DESeq2 results object</span></span>
<span><span class="co"># Filter genes that have nonzero counts</span></span>
<span><span class="va">bg_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">dds_macrophage</span><span class="op">)</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">dds_macrophage</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load required package for analysis</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"topGO"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform Gene Ontology enrichment analysis using the topGOtable function from </span></span>
<span><span class="co"># the "pcaExplorer" package</span></span>
<span><span class="va">macrophage_topGO_example</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">pcaExplorer</span><span class="fu">::</span><span class="fu"><a href="https://federicomarini.github.io/pcaExplorer/reference/topGOtable.html" class="external-link">topGOtable</a></span><span class="op">(</span><span class="va">de_symbols_IFNg_vs_naive</span>,</span>
<span>    <span class="va">bg_ids</span>,</span>
<span>    ontology <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>    mapping <span class="op">=</span> <span class="st">"org.Hs.eg.db"</span>,</span>
<span>    geneID <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>    topTablerows <span class="op">=</span> <span class="fl">500</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>As mentioned earlier, <code>GeDi</code> expects the input to contain
at least two columns: one named “Genesets” and one named “Genes”. While
this is not strictly mandatory when providing your data interactively
during an application session, it becomes necessary if you intend to
initiate the application with your input as parameters (e.g.,
<code>GeDi(genesets = my_genesets_df)</code>). In such cases, the
“Genesets” column should contain identifiers for each geneset in the
input, while the “Genes” column should consist of comma-separated lists
of genes associated with each geneset.</p>
<p>Therefore, we will adjust the column names of the resulting
<code>DataFrame</code> from the enrichment analysis to adhere to the
required format.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Rename columns in the macrophage_topGO_example dataframe</span></span>
<span><span class="co"># Change the column name "GO.ID" to "Genesets"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">macrophage_topGO_example</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">macrophage_topGO_example</span><span class="op">)</span> <span class="op">==</span> <span class="st">"GO.ID"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Genesets"</span></span>
<span></span>
<span><span class="co"># Change the column name "genes" to "Genes"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">macrophage_topGO_example</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">macrophage_topGO_example</span><span class="op">)</span> <span class="op">==</span> <span class="st">"genes"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Genes"</span></span></code></pre></div>
<div class="section level3">
<h3 id="all-set">All set!<a class="anchor" aria-label="anchor" href="#all-set"></a>
</h3>
<p>Now that we’ve obtained functional annotation results from the <em><a href="https://bioconductor.org/packages/3.19/macrophage" class="external-link">macrophage</a></em>
dataset, we can begin exploring the data using <code>GeDi</code>. You
have two options: you can either launch the application and supply the
generated data using the <code><a href="../reference/GeDi.html">GeDi()</a></code> command, or if you’ve
followed this vignette, you can initiate the application directly with
the loaded data by executing
<code>GeDi(genesets = macrophage_topGO_example)</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GeDi.html">GeDi</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/GeDi.html">GeDi</a></span><span class="op">(</span>genesets <span class="op">=</span> <span class="va">macrophage_topGO_example</span><span class="op">)</span></span></code></pre></div>
<p>The above shown code will open the application, directing you to the
<strong>Welcome</strong> page. The <strong>Welcome</strong> page of
<code>GeDi</code> serves as the entry point to the application,
providing users with an overview of its features and functionalities.
Upon launching the application, users are greeted with a user-friendly
interface designed to facilitate the exploration and interpretation of
functional annotation and enrichment analysis results. The
<strong>Welcome</strong> page offers guidance on how to navigate the
application and highlights key components such as data input options,
visualization tools, and interactive features. Whether users are new to
GeDi or returning to explore additional datasets, the
<strong>Welcome</strong> page serves as a central hub for accessing
resources and getting started with their analysis journey.</p>
<div class="figure" style="text-align: center">
<img src="Welcome_page.png" alt="The Welcome panel of GeDi" width="1453"><p class="caption">
The Welcome panel of GeDi
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="userinterface">Description of the <code>GeDi</code> user interface<a class="anchor" aria-label="anchor" href="#userinterface"></a>
</h2>
<p>The <code>GeDi</code> application, developed with the <em><a href="https://CRAN.R-project.org/package=shiny" class="external-link">shiny</a></em>
framework, incorporates the modern design elements of the <em><a href="https://CRAN.R-project.org/package=bs4Dash" class="external-link">bs4Dash</a></em>
package, which is built upon Bootstrap 4. This combination of
technologies ensures a sleek and visually appealing user interface for
navigating and interacting with the functionality offered by
<code>GeDi</code>. By leveraging the features of <em><a href="https://CRAN.R-project.org/package=shiny" class="external-link">shiny</a></em> and
<em><a href="https://CRAN.R-project.org/package=bs4Dash" class="external-link">bs4Dash</a></em>,
<code>GeDi</code> provides users with an intuitive and aesthetically
pleasing environment for conducting functional annotation and enrichment
analyses on their datasets.</p>
<div class="section level3">
<h3 id="header-navbar">Header (navbar)<a class="anchor" aria-label="anchor" href="#header-navbar"></a>
</h3>
<p>The dashboard navbar in <code>GeDi</code>, referred to as such in the
<em><a href="https://CRAN.R-project.org/package=bs4Dash" class="external-link">bs4Dash</a></em>
framework, features two dropdown menus accessible by clicking on their
respective icons (<i class="fa fa-book"></i> and
<i class="fa fa-info"></i>). These menus offer additional functionality
through various buttons:</p>
<ul>
<li>
<i class="fa fa-book"></i> - This option allows users to explore the
<code>GeDi</code> vignette, either the version bundled with the package
or the online version, providing detailed documentation and usage
guidelines.</li>
<li>
<i class="fa fa-question-circle"></i> - Clicking this icon provides
users with initial help information about <code>GeDi</code>, offering a
brief overview of the package’s functionality and outlining the
prerequisites for proper execution.</li>
<li>
<i class="fa fa-info-circle"></i> - Selecting this option displays
information about the current session, presenting details such as the R
environment and loaded packages, helpful for troubleshooting and
debugging purposes.</li>
<li>
<i class="fa fa-heart"></i> - This button offers general information
about <code>GeDi</code>, including links to its development version for
contribution and guidelines on citing the tool in research
publications.</li>
</ul>
<p>Besides the two dropdown menus, users can also find the
<code>Bookmark</code> button in the Navbar. The <code>Bookmark</code>
button in the <code>GeDi</code> navbar serves as a convenient tool for
users to save and bookmark genes and genesets of interest for later
reference. To use this feature, users must first select or click on a
gene or geneset that they wish to bookmark. Once the desired gene or
geneset is selected, users can then click on the <code>Bookmark</code>
button to add it to a list of bookmarked items within the
<code>GeDi</code> application. This functionality enables users to
organize and revisit specific genes or genesets that they find relevant
or intriguing during their exploration of functional annotation and
enrichment analysis results. The bookmarked genes and genesets can later
be found in the <strong>Report</strong> panel.</p>
</div>
<div class="section level3">
<h3 id="sidebar">Sidebar<a class="anchor" aria-label="anchor" href="#sidebar"></a>
</h3>
<p>By clicking the <i class="fa fa-bars"></i> menu bar icon on the left
side of the app (or simply by moving the mouse over to the left side if
viewing the app in full screen mode), users can activate the sidebar
menu. This sidebar menu serves as the primary means of accessing the
various panels of the <code>GeDi</code> application, providing
navigation to different functionalities. More detailed explanations of
each panel will be provided in the next section.</p>
</div>
<div class="section level3">
<h3 id="body">Body<a class="anchor" aria-label="anchor" href="#body"></a>
</h3>
<p>The structure of <code>GeDi</code> is designed around different
panels, each of which becomes active upon clicking the corresponding
icons or text in the sidebar.</p>
<p>While the Welcome panel is relatively self-explanatory, additional
information and explanations are provided for the functionality of the
remaining panels. For new users seeking guidance, there’s a question
circle button (<i class="fa fa-question-circle"></i>) available to
initiate an interactive tour of <code>GeDi</code>. This tour allows
users to learn the basic usage mechanisms by actively engaging with the
interface. During the tour, specific elements are highlighted in
response to user actions, while the rest of the UI remains shaded to
maintain focus. Users can interrupt the tour at any time by clicking
outside the highlighted window, and navigation between steps is
facilitated by arrow buttons (left, right). The tour functionality is
implemented using the <em><a href="https://CRAN.R-project.org/package=rintrojs" class="external-link">rintrojs</a></em>
package.</p>
</div>
</div>
<div class="section level2">
<h2 id="functionality">The <code>GeDi</code> functionality<a class="anchor" aria-label="anchor" href="#functionality"></a>
</h2>
<p>The <code>GeDi</code> <em><a href="https://bioconductor.org/packages/3.19/shiny" class="external-link">shiny</a></em>
application is organized into distinct panels, each serving a specific
purpose, which will be thoroughly explored in the following
sections.</p>
<div class="section level3">
<h3 id="the-welcome-panel">The Welcome panel<a class="anchor" aria-label="anchor" href="#the-welcome-panel"></a>
</h3>
<p>This panel serves as a guide for utilizing <code>GeDi</code>
effectively. It offers detailed instructions on generating input data
for the application, elucidating the expected input format and outlining
the various interactive elements present in the app’s other panels.</p>
<div class="figure" style="text-align: center">
<img src="Welcome_page.png" alt="The Welcome panel of GeDi" width="1453"><p class="caption">
The Welcome panel of GeDi
</p>
</div>
</div>
<div class="section level3">
<h3 id="the-data-input-panel">The Data Input panel<a class="anchor" aria-label="anchor" href="#the-data-input-panel"></a>
</h3>
<p>This panel serves as a hub for managing data input if it’s not
provided within the function call. It’s divided into distinct boxes,
each representing a step of the data input process, which sequentially
appear as you successfully complete each preceding step.</p>
<p><strong>Step 1</strong>: Provide your Genesets as input data</p>
<p>In the initial <strong>Step 1</strong> box, you can provide your data
by utilizing the <strong>Browse</strong> button. This action opens a
modal window enabling you to select the relevant file from your computer
storage. After successfully loading the data, a preview is displayed in
the <strong>Genesets preview</strong> box on the right. During this
step, the application checks if your input contains the “Genesets” and
“Genes” columns. If these columns are missing, a small error message
appears in the lower right corner. Additionally, two drop-down menus
allow you to select the correct columns from your data and update the
input accordingly.</p>
<p>You also have the option to start using <code>GeDi</code> with
preprocessed example data based on the <em><a href="https://bioconductor.org/packages/3.19/macrophage" class="external-link">macrophage</a></em>
dataset. Simply click the <strong>Load demo data</strong> button to load
the example data’s enrichment results. You can explore these results in
the <strong>Genesets preview</strong> box.</p>
<p>However, instead of loading demo data and observing the expected data
structure through the <strong>Genesets preview</strong> box, you can
also use the <strong>Have a look at the data structure</strong> button.
By clicking this button, a modal window with a visual representation of
the expected input data structure will open. This screenshot serves as a
helpful guide, providing you with a clear understanding of how your data
should be formatted for optimal compatibility with
<code>GeDi</code>.</p>
<p>Once, you have successfully loaded some data, the data input process
will proceed and two additional boxes will be displayed in the
panel.</p>
<div class="figure" style="text-align: center">
<img src="Data_Input_panel_Step1.png" alt="The Data input panel - Step 1" width="1512"><p class="caption">
The Data input panel - Step 1
</p>
</div>
<p><strong>Optional Filtering Step</strong>: Filter generic genesets</p>
<p>Introducing the first new box, the <strong>Optional Filtering
Step</strong> offers a non-compulsory yet advantageous opportunity to
refine your geneset selection. While not obligatory for data
exploration, engaging in this step can notably optimize downstream
processing runtime. Here, you’re empowered to filter genesets within
your dataset, thereby enhancing result interpretation. This step enables
the exclusion of large and generic genesets, contributing to clearer
insights. Additionally, you have the flexibility to filter genesets
based on size criteria.</p>
<p>The box features a histogram illustrating geneset sizes, providing
visual context for the filtering process. Within the interface, two
input fields are available for customization. The left input field
facilitates the selection of individual genesets by their identifiers in
the “Genesets” column of your dataset. Meanwhile, the right input field
empowers you to establish a threshold “x” for filtering genesets with a
size greater than or equal to “x.” This interactive approach ensures
tailored filtering suited to your specific analysis requirements.</p>
<p>Once you’ve chosen the genesets you wish to exclude from your
dataset, you can initiate the filtering process by clicking the “Remove
the selected Genesets” button. This action will remove all selected
genesets from the dataset. Additionally, you have the option to save the
filtered data using the “Download the filtered data” button. Clicking
this button will save the filtered data to your local machine. This
feature can be particularly beneficial for users who intend to revisit
their data in a new instance of GeDi and want to ensure that previously
identified uninsightful genesets have already been filtered out.</p>
<p>Once you’ve chosen the genesets you wish to exclude from your
dataset, you can initiate the filtering process by clicking the
<code>Remove the selected Genesets</code> button. This action will
remove all selected genesets from the dataset. Additionally, you have
the option to save the filtered data using the
<code>Download the filtered data</code> button. Clicking this button
will save the filtered data to your local machine.This feature can be
particularly beneficial for users who intend to revisit their data in a
new instance of <code>GeDi</code> and want to ensure that previously
identified uninsightful genesets have already been filtered out.</p>
<div class="figure" style="text-align: center">
<img src="Optional_Filtering.png" alt="Optional Filtering Step" width="1512"><p class="caption">
Optional Filtering Step
</p>
</div>
<p><strong>Step 2</strong>: Species Selection</p>
<p>Upon advancing to the second box labeled <strong>Step 2</strong>,
you’ll encounter the crucial task of selecting the species associated
with your dataset. This step holds significant importance for the
computation of the <strong>PMM score</strong> within <code>GeDi</code>,
which heavily relies on a <strong>Protein-Protein Interaction
(PPI)</strong> matrix. This matrix plays a pivotal role in capturing
protein interaction strength, thereby enriching distance scores with
valuable biological context. To access and utilize this essential
information, specifying the species linked to your dataset is mandatory.
By clicking the input field, you’ll prompt a dropdown menu showcasing
preselected species options. If your species is included, simply make
your selection. Alternatively, if your species is not listed, you have
the option to manually input it. In cases of uncertainty, a convenient
link provided on the right directs you to the STRING database, enabling
verification of species details and PPI availability for informed
decision-making.</p>
<div class="figure" style="text-align: center">
<img src="Species_Selection.png" alt="Species Selection" width="1512"><p class="caption">
Species Selection
</p>
</div>
<p><strong>Step 3</strong>: PPI Matrix Download</p>
<p>Following species selection, a third box named <strong>Step
3</strong> will emerge. In this phase, you have the opportunity to
download the Protein-Protein Interaction (PPI) matrix. This process may
necessitate some time, with a progress bar positioned in the lower right
corner providing real-time updates on the download status. Once the
download is complete, you can conveniently preview the PPI matrix within
the <strong>PPI Preview</strong> box situated on the right-hand side of
the interface. This will show that the PPI consists of three columns:
<strong>Gene1</strong> and <strong>Gene2</strong>, housing the gene
symbols corresponding to the interacting proteins, and a column labeled
<strong>combined_score</strong>, denoting the confidence level of each
interaction. The assigned score is derived from the number of known
interactions between two proteins, normalized to the (0, 1) interval
utilizing the formula:</p>
<p><span class="math display">\[
\begin{aligned}
combinedScore = \frac{(\#interaction - min)}{(max - min)}
\end{aligned}
\]</span></p>
<p>where <strong>min</strong> and <strong>max</strong> represent the
minimum and maximum number of interactions, respectively.</p>
<p>In addition to downloading a PPI matrix during the current session,
users can also upload a previously saved matrix for analysis using the
<strong>Browse</strong> button. This functionality allows users to work
with their own customized datasets or previously analyzed PPI matrices.
Furthermore, saving the downloaded PPI matrix locally enables users to
store the data on their machine for future use. By saving the matrix
locally via the <strong>Save PPI matrix</strong> button, users can
access the data quickly in subsequent sessions without having to wait
for the download process again. This capability significantly enhances
workflow efficiency and allows for seamless continuation of analysis
across different sessions.</p>
<div class="figure" style="text-align: center">
<img src="Downloading_PPI.png" alt="Downloading the PPI" width="1495"><p class="caption">
Downloading the PPI
</p>
</div>
<p>While the final two steps are optional, note that the PPI matrix is
only required for a singular score. Therefore, you can commence data
exploration without necessarily completing these additional steps.</p>
<p>Upon concluding the essential tasks outlined in this panel, you are
ready to progress to the <strong>Distance Scores</strong> panel.</p>
</div>
<div class="section level3">
<h3 id="the-distance-scores-panel">The Distance Scores panel<a class="anchor" aria-label="anchor" href="#the-distance-scores-panel"></a>
</h3>
<p>This panel focuses mainly on computing distance scores for the
provided input data. Like the preceding panel, it is segmented into two
distinct sections, each serving a specific function.</p>
<div class="figure" style="text-align: center">
<img src="Distance_Score_panel.png" alt="The Distance Score panel" width="1512"><p class="caption">
The Distance Score panel
</p>
</div>
<p><strong>Calculating Distance Scores</strong></p>
<p>In the upper box, titled <strong>Calculate distance scores for your
Genesets</strong>, you have the flexibility to select from various
distance scores for computation. This feature provides users with a
range of options to tailor the analysis according to their specific
requirements and preferences. The available scores are:</p>
<ul>
<li>
<strong>PMM Score</strong>: This score integrates protein-protein
interaction (PPI) data into the Meet-Min distance. The PPI-weighted
Meet-Min (<strong>PMM</strong>) score is defined as</li>
</ul>
<p><span class="math display">\[
\begin{aligned}
PMM = min(PMM(A -&gt; B), PMM(B -&gt; A))
\end{aligned}
\]</span> where</p>
<p><span class="math display">\[
\begin{aligned}
PMM(X -&gt; Y) = 1 - \frac{|X \cap Y|}{min(|X|, |Y|)} - \\
\frac{\alpha}{min(|X|,|Y|)} \sum_{x \in X-Y} \frac{w \sum_{y \in X \cap
Y} PPI(x, y) + \sum_{y \in Y - X}PPI(x, y)}{max(PPI) (w |X \cap Y| + |Y
- X|)}
\end{aligned}
\]</span> and</p>
<p><span class="math display">\[
\begin{aligned}
w = \frac{min(|X|, |Y|)}{|X| + |Y|}
\end{aligned}
\]</span> <span class="math inline">\(\alpha\)</span> is a scaling
factor between 0 and 1. The PPI matrix can be downloaded from the
<strong>Data Input</strong> panel. More details can be found in the
paper by Yoon et al <span class="citation">(Yoon et al.
2019)</span>.</p>
<ul>
<li>
<strong>Kappa Score</strong>: The <strong>Kappa</strong> distance is
a set-based metric based on observed and expected agreement rates
between two genesets. It is defined as</li>
</ul>
<p><span class="math display">\[
\begin{aligned}
Kappa = 1 - \frac{O - E}{1 - E}
\end{aligned}
\]</span></p>
<p>where</p>
<p><span class="math display">\[
\begin{aligned}
O = \frac{|A \cap B| + |A \cup B|^c}{U} \\
E = \frac{|A| |B| + |A^c| |B^c|}{|U|^2}
\end{aligned}
\]</span></p>
<p>U is the set of all unique genes in the data. In this application the
Kappa distance is additionally normalized to the (0, 1) interval to make
it comparable to the remaining distance metrics.</p>
<ul>
<li>
<strong>Jaccard Score</strong>: The <strong>Jaccard</strong>
distance uses the Jaccard coefficient, which is transformed into a
distance metric by subtracting it from 1. It is defined as</li>
</ul>
<p><span class="math display">\[
\begin{aligned}
Jaccard = 1 - \frac{|A \cap B|}{|A \cup B|}
\end{aligned}
\]</span></p>
<ul>
<li>
<strong>Meet-Min Score</strong>: The <strong>Meet-Min</strong> (MM)
distance transforms the overlap coefficient into a distance measure by
subtracting it from 1.The overlap coefficient is a similarity measure
which is defined as</li>
</ul>
<p><span class="math display">\[
\begin{aligned}
OC = \frac{|A \cap B|}{min(|A|, |B|)}
\end{aligned}
\]</span></p>
<p>In order to transform this measure of similarity into a measure of
distance, the overlap coefficient is subtracted from 1, resulting in the
calculation of the Meet-Min (MM) distance as</p>
<p><span class="math display">\[
\begin{aligned}
MM = 1 - \frac{|A \cap B|}{min(|A|, |B|)}
\end{aligned}
\]</span></p>
<p>As a solely set based measurement, the Meet-Min distance only takes
the composition of the genesets into account but not the underlying
biological information inherent in the genesets.</p>
<p>Each scoring method possesses its own set of advantages and
drawbacks, underscoring the importance of selecting one that suits your
dataset characteristics and analysis goals. Upon choosing a score, the
<strong>Compute the distances between genesets</strong> button appears
on the on the right side. Clicking this button initiates the scoring
procedure, which may require some time to execute, particularly for
larger datasets. To monitor the progress of this operation, refer to the
progress bar located in the lower right corner of the panel. Once the
scoring process concludes, you can delve into the <strong>Geneset
Distance Scores</strong> box to explore a variety of visual
representations of your data.</p>
<p><strong>Distance Scores Visualizations</strong></p>
<ul>
<li><p><strong>Distance Scores Heatmap</strong>: The initial
visualization offered is a heatmap illustrating the distribution of
distance scores. Activation of the heatmap generation is triggered by
clicking the <strong>Calculate Distance Score Heatmap</strong> button.
Following computation, users can interact with the heatmap by hovering
over it, revealing the involved genesets and their corresponding scores.
Additionally, users can zoom in on specific areas of interest. To reset
the zoomed view, a simple click outside the heatmap area
suffices.</p></li>
<li><p><strong>Distance Scores Dendrogram</strong>: The second
visualization provided is a dendrogram showcasing individual distance
scores. Hierarchical clustering is employed to generate the dendrogram,
which effectively groups genesets exhibiting the highest similarity. To
enhance the dendrogram’s presentation, users can select different
combination methods using the drop-down menu located on the left
side.</p></li>
<li><p><strong>Distance Scores Graph</strong>: The final visualization
available is the network representation of distance scores. In this
representation, nodes/genesets with scores below a predefined threshold
are connected by edges. By default, the threshold is set to 0.3, but
users can adjust it via the slider located on the left. This interactive
graph allows users to hover over or click on nodes to highlight
connected nodes and obtain additional information about genesets upon
selection. Furthermore, users can search for specific genesets using the
input field on the left, with the selected geneset being subsequently
highlighted in the graph. The <strong>Graph metrics</strong> table at
the bottom of this box contains various metrics pertaining to the graph,
such as degree, betweenness, harmonic centrality, clustering
coefficient, and input data. This tabulated information serves to
provide users with valuable insights into the underlying data and
distance scores.</p></li>
</ul>
<p><strong>Bookmarking from the this panel:</strong></p>
<p>As users navigate through the distance scores of genesets in this
section, they may encounter genesets and interactions that capture their
interest and merit further investigation. To aid in preserving these
noteworthy genesets for later exploration, you can utilize the
<strong>Bookmark</strong> button situated in the Navbar. Upon clicking
this button, the selected geneset will be added to the list of
bookmarked genesets within the <strong>Report</strong> panel.
Additionally, informative messages displayed in the lower right corner
will guide users through the bookmarking process.</p>
<p>Once you’ve finished exploring the distance scores, you can proceed
to the <strong>Clustering graph</strong> panel.</p>
</div>
<div class="section level3">
<h3 id="the-clustering-graph-panel">The Clustering Graph panel<a class="anchor" aria-label="anchor" href="#the-clustering-graph-panel"></a>
</h3>
<p>This panel is dedicated to the computation of clusters among genesets
based on their similarity, which is derived from the previously
calculated distance scores. Similar to the preceding panel, it comprises
two distinct boxes. Within these boxes, users can access functionalities
to determine and visualize clusters of genesets that exhibit comparable
characteristics or functions.</p>
<p>The computation of clusters involves grouping genesets that display
similar patterns of distance scores, thereby indicating shared
biological characteristics or functional relationships. This clustering
process enables users to identify cohesive groups of genesets with
related functionalities or involvement in similar biological
processes.</p>
<div class="figure" style="text-align: center">
<img src="Clustering_Graph_Panel.png" alt="The Clustering Graph panel" width="1512"><p class="caption">
The Clustering Graph panel
</p>
</div>
<p><strong>Choosing a Clustering algorithm</strong></p>
<p>The upper box, labeled <strong>Select the clustering method</strong>,
provides a selection of distinct clustering algorithms. Users can
explore various options to find the most suitable algorithm for their
analysis:</p>
<ul>
<li><p><strong>Louvain</strong>: The Louvain algorithm, a prevalent tool
in biological network analysis, seeks to divide graph nodes into
clusters to optimize the modularity metric. This metric gauges the
strength of connections within clusters relative to those between
clusters. Consequently, nodes within the same cluster exhibit greater
similarity to one another than to nodes outside the cluster. This
clustering approach aims to enhance data interpretation by grouping
similar genesets together. Users can adjust a slider in the bottom left
corner of the box to set a similarity threshold, determining when
genesets are considered similar based on distance scores.</p></li>
<li><p><strong>Markov</strong>: The Markov algorithm, commonly employed
in biological network analysis, is designed to pinpoint densely
interconnected regions within graphs. These regions frequently align
with communities or clusters in the graph structure. Users can utilize a
slider located in the bottom left corner of the box to specify a
similarity threshold, determining when genesets are deemed similar based
on distance scores.</p></li>
<li><p><strong>Fuzzy clustering</strong>: The Fuzzy Clustering algorithm
is a computational technique used to partition data points into clusters
based on their similarity, while allowing for data points to belong to
multiple clusters with varying degrees of membership. It operates
through distinct steps and requires the specification of different
thresholds. Firstly, the <strong>Similarity threshold</strong> is set to
determine if two genesets exhibit sufficient similarity to be
potentially clustered together. Secondly, the <strong>Membership
threshold</strong> dictates how many members of a potential cluster must
possess a close relationship, defined by a distance score less than or
equal to the similarity threshold, for the cluster to persist. Lastly,
the <strong>Clustering threshold</strong> determines whether two
clusters will be merged. Clusters are merged if their percentage of
overlap meets or exceeds the clustering threshold. Users can adjust all
thresholds using sliders provided in the interface.</p></li>
<li><p><strong>kNN</strong>: The k-nearest neighbor clustering algorithm
groups nodes with their k nearest neighbors, where <strong>k</strong> is
a parameter defined by the user. Through this algorithm, each node is
assigned to the cluster containing its k nearest neighbors based on the
calculated distance scores. The user can specify the value of k using a
slider provided in the interface, enabling customization of the
clustering process according to the specific requirements of the
analysis. Adjusting the value of k allows for exploration of different
levels of granularity in the clustering results, providing flexibility
in the interpretation of the data.</p></li>
</ul>
<p>Once you choose a method, you can start the cluster calculation via
the <strong>Cluster the Genesets</strong> button on the right. Keep in
mind that this step might take some time, especially for larger
datasets. Look for the progress bar in the lower right corner for
updates on the scoring status.</p>
<p>Once the clusters are calculated, you can explore various
visualizations of your data in the <strong>Geneset Cluster
Graphs</strong> box.</p>
<p><strong>Cluster Visualizations</strong></p>
<ul>
<li><p><strong>Geneset Graph</strong>: In the <strong>Geneset
Graph</strong>, clusters are visualized as a graph, with individual
genesets serving as nodes and edges connecting genesets within the same
cluster. To highlight specific nodes, utilize the <strong>Select by
id</strong> feature on the left, or choose to highlight entire clusters
by selecting the respective option under <strong>Select by
cluster</strong>. Please note that only genesets belonging to at least
one cluster will be displayed in this graph. For additional insights,
nodes can be colored based on specific parameters from your input data,
accessible through the <strong>Color the graph</strong> by dropdown
menu. Depending on the information provided with your data, various
options will be available. While interacting with the network, nodes can
be moved by clicking and dragging them to desired locations, offering
flexibility in managing node placement which is particularly useful in
complex or densely populated graphs.</p></li>
<li><p><strong>Cluster-Geneset Bipartite Graph</strong>: The
<strong>Cluster-Geneset Bipartite Graph</strong> presents a bipartite
representation of the clusters. In this visualization, nodes represent
both clusters and genesets, with edges connecting cluster nodes to their
corresponding geneset members. Hovering over nodes provides additional
data insights. Cluster nodes display the members within each cluster,
while geneset nodes showcase the genes associated with each
geneset.</p></li>
<li><p><strong>Cluster Enrichment Terms Word Cloud</strong>: The
<strong>Cluster Enrichment Terms Word Cloud</strong> displays the most
frequently occurring terms for each cluster. This visualization proves
particularly useful when your data includes brief descriptions of the
genesets, in addition to the mandatory input data. By utilizing the
<strong>Select a cluster</strong> drop-down menu, you can designate the
cluster of interest. Furthermore, hovering over the word cloud enables
you to select individual terms and view the frequency with which each
term appears in the descriptions of the genesets within that
cluster.</p></li>
<li><p><strong>Clustering graph summaries</strong>: The cluster
information is also summarized in a table-like format in the
<strong>Clustering graph summaries</strong> box. This table displays
each geneset alongside the cluster to which it belongs. Additionally,
the table features a search function, facilitating the quick retrieval
of a geneset of interest.</p></li>
</ul>
<p><strong>Bookmarking from this panel:</strong></p>
<p>While exploring the <strong>Clustering Graph panel</strong>, users
may encounter genesets and clusters that intrigue them and warrant
further investigation. To facilitate the preservation of these notable
genesets and clusters for future exploration, users can utilize the
<strong>Bookmark</strong> button located in the Navbar. Clicking this
button will add the selected geneset or cluster to the list of
bookmarked items within the Report panel. Helpful messages displayed in
the lower right corner will assist users throughout the bookmarking
process.</p>
<p>In order to bookmark interesting genes and clusters, users simply
select a geneset or cluster from the Geneset Graph or Cluster-Geneset
Bipartite Graph and use the Bookmark button to add the respective
information to the set of bookmarked features. After exploring the
results in the <strong>Clustering Graph panel</strong>, users can
proceed to the <strong>Report</strong> panel to have a look at the
bookmarked genesets and clusters or iterate through the individual
panels of the app for a more in depth exploration of the data.</p>
</div>
<div class="section level3">
<h3 id="the-report-panel">The Report panel<a class="anchor" aria-label="anchor" href="#the-report-panel"></a>
</h3>
<p>In this panel of the application, users can obtain a comprehensive
overview of the items they have bookmarked for further exploration. On
the left side of the interface, bookmarked genesets are listed, while
bookmarked clusters are displayed on the right side.</p>
<p>During an interactive exploration session, recalling specific details
about each bookmarked item can sometimes be challenging. Therefore,
users are provided with convenient options to manage their bookmarked
data.</p>
<p>Below the interactive tables displaying bookmarked genesets and
clusters, users can find buttons allowing them to download the content
of each table individually. Additionally, the <strong>Start the
generation of the report</strong> button is provided to generate a
detailed report encompassing all selected elements of interest.</p>
<p>The report generation process utilizes a predefined template report
included within the <code>GeDi</code> package. This template leverages
the input elements and reactive values associated with the bookmarks,
ensuring that the generated report contains comprehensive and relevant
information.</p>
<p>The resulting report serves as a valuable tool for creating a
permanent and reproducible analysis output. Users can easily store or
share this report for future reference or collaboration purposes.</p>
<div class="figure" style="text-align: center">
<img src="Report_panel.png" alt="The Report panel" width="1512"><p class="caption">
The Report panel
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="additionalinfo">Additional Information<a class="anchor" aria-label="anchor" href="#additionalinfo"></a>
</h2>
<p>Bug reports can be submitted on the Bioconductor <a href="https://support.bioconductor.org/" class="external-link">support site</a> or opened as
issues in the <code>GeDi</code> <a href="https://github.com/AnnekathrinSilvia/GeDi/issues" class="external-link">GitHub
repository</a>. Please note that the GitHub repository also hosts the
development version of the package, where new functionality is
continuously added - be cautious, as you may be working with
cutting-edge versions!</p>
<p>The authors welcome thoughtful suggestions for enhancements or new
features, and even better, pull requests.</p>
<p>If you utilize <code>GeDi</code> for your analysis, kindly cite it
using the provided citation format below:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"GeDi"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="additional-example-data">Additional example data<a class="anchor" aria-label="anchor" href="#additional-example-data"></a>
</h2>
<p>In this section, we present additional examples demonstrating the
versatility of <code>GeDi</code> in analyzing functional enrichment data
from two widely used databases, KEGG and Reactome. By leveraging the
rich resources provided by these databases, GeDi offers researchers a
comprehensive toolkit for exploring and interpreting complex biological
pathways and processes. Through step-by-step demonstrations, we
illustrate how GeDi can seamlessly integrate with data from KEGG and
Reactome, enabling users to gain deeper insights into the functional
annotations of their gene sets. Whether investigating specific pathways
or broader biological processes, GeDi provides intuitive and powerful
functionalities to enhance the analysis of functional enrichment data
from diverse sources.</p>
<p>In this section we will demonstrate how results containing
identifiers from databases like KEGG <span class="citation">(Kanehisa et
al. 2023)</span> or Reactome <span class="citation">(Gillespie et al.
2022)</span> - e.g. generated using <code>enrichKegg</code> or
<code>enrichPathway</code> functions from the <em><a href="https://bioconductor.org/packages/3.19/clusterProfiler" class="external-link">clusterProfiler</a></em>
package - can be utilized as input for <code>GeDi</code>. We will again
use the data of the <em><a href="https://bioconductor.org/packages/3.19/macrophage" class="external-link">macrophage</a></em>
package, specifically the differentially expressed genes we have
identified before. With this data, we demonstrate how to generate the
results and prepare them for their use in <code>GeDi</code>.</p>
<p>However, before we can use the <code>enrichKEGG</code> function from
the <em><a href="https://bioconductor.org/packages/3.19/clusterProfiler" class="external-link">clusterProfiler</a></em>
package, we have to map the ENSEMBL ids of the data to Entrez ids. For
this, we will up the first use the <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
package to generate a mapping of ENSEMBL to Entrez.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the "biomaRt" package to access the BioMart database</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/grimbough/biomaRt" class="external-link">"biomaRt"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set up a connection to the ENSEMBL BioMart database for human genes</span></span>
<span><span class="va">mart</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html" class="external-link">useMart</a></span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>, dataset <span class="op">=</span> <span class="st">"hsapiens_gene_ensembl"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Retrieve gene annotations using the BioMart database</span></span>
<span><span class="va">anns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html" class="external-link">getBM</a></span><span class="op">(</span></span>
<span>  attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"ensembl_gene_id"</span>,</span>
<span>    <span class="st">"external_gene_name"</span>,</span>
<span>    <span class="st">"entrezgene_id"</span>,</span>
<span>    <span class="st">"description"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  filters <span class="op">=</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span>  values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dds_macrophage</span><span class="op">)</span>,</span>
<span>  mart <span class="op">=</span> <span class="va">mart</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Match the retrieved annotations to the genes in dds_macrophage</span></span>
<span><span class="va">anns</span> <span class="op">&lt;-</span> <span class="va">anns</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dds_macrophage</span><span class="op">)</span>, <span class="va">anns</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
<p>Next, we map the differentially expressed genes to get the right
identifiers and run the <code>enrichKEGG</code> function. We set the
organism to human and the p-value cutoff to 5%.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the "clusterProfiler" package for functional enrichment analysis</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">"clusterProfiler"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Retrieve Entrez gene IDs from the annotations data frame based on matching </span></span>
<span><span class="co"># Ensembl gene IDs from the DE results</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="va">anns</span><span class="op">$</span><span class="va">entrezgene_id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res_macrophage_IFNg_vs_naive</span><span class="op">)</span>, <span class="va">anns</span><span class="op">$</span><span class="va">ensembl_gene_id</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Perform KEGG pathway enrichment analysis using the retrieved gene IDs</span></span>
<span><span class="va">res_enrich</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichKEGG.html" class="external-link">enrichKEGG</a></span><span class="op">(</span><span class="va">genes</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"hsa"</span>,</span>
<span>  pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can now use the results of the enrichment in <code>GeDi</code>.
For this, we directly start the app with the loaded data. If you have
not computed the data following this workflow, you can beforehand load
it from the available data in this package.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the "macrophage_KEGG_example" dataset from the "GeDi" package</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"macrophage_KEGG_example"</span>, package <span class="op">=</span> <span class="st">"GeDi"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Start the GeDi app with the loaded data</span></span>
<span><span class="co"># The "genesets" parameter is set to the loaded "macrophage_KEGG_example" dataset</span></span>
<span><span class="fu"><a href="../reference/GeDi.html">GeDi</a></span><span class="op">(</span>genesets <span class="op">=</span> <span class="va">macrophage_KEGG_example</span><span class="op">)</span></span></code></pre></div>
<p>In a similar manner we can use the Reactome database for the
functional annotation. Here, we use the <em><a href="https://bioconductor.org/packages/3.19/ReactomePA" class="external-link">ReactomePA</a></em>
package and the differentially expressed genes.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the "ReactomePA" package for pathway enrichment analysis</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">"ReactomePA"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform pathway enrichment analysis using the "enrichPathway" function</span></span>
<span><span class="va">reactome</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReactomePA/man/enrichPathway.html" class="external-link">enrichPathway</a></span><span class="op">(</span><span class="va">genes</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"human"</span>,</span>
<span>  pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  readable <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now we can use the results in the same manner as for the KEGG pathway
analysis.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the "macrophage_Reactome_example" dataset from the "GeDi" package</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"macrophage_Reactome_example"</span>, package <span class="op">=</span> <span class="st">"GeDi"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Start the GeDi app with the loaded data</span></span>
<span><span class="co"># The "genesets" parameter is set to the loaded "macrophage_Reactome_example" dataset</span></span>
<span><span class="fu"><a href="../reference/GeDi.html">GeDi</a></span><span class="op">(</span>genesets <span class="op">=</span> <span class="va">macrophage_Reactome_example</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="faqs">FAQs<a class="anchor" aria-label="anchor" href="#faqs"></a>
</h2>
<p><strong>Q: My configuration on two machines is somewhat different, so
I am having difficulty in finding out what packages are different. Is
there something to help on this?</strong></p>
<p>A: Yes, you can check out <em><a href="https://github.com/federicomarini/sessionDiffo" class="external-link">sessionDiffo</a></em>,
a small utility to compare the outputs of two different
<code>sessionInfo</code> outputs. This can help you pinpoint what
packages might be causing the issue.</p>
<p><strong>Q: I am using a different service/software for generating the
results of functional enrichment analysis. How do I plug this into
<code>GeDi</code>?</strong></p>
<p>A: You can use nearly any result of a functional enrichment analysis
in <code>GeDi</code> as long as the results are transformed in a way
that they fit the input requirements. Please check out the
<strong>Welcome</strong> page to see the specification of the input
requirements.</p>
</div>
<div class="section level2">
<h2 class="unnumbered smaller" id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R Under development (unstable) (2024-03-12 r86109)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Monterey 12.7.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Berlin</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] org.Hs.eg.db_3.18.0         topGO_2.55.0                SparseM_1.81               </span></span>
<span><span class="co">#&gt;  [4] GO.db_3.18.0                graph_1.81.0                AnnotationDbi_1.65.2       </span></span>
<span><span class="co">#&gt;  [7] GeneTonic_2.7.2             pcaExplorer_2.29.0          bigmemory_4.6.4            </span></span>
<span><span class="co">#&gt; [10] DESeq2_1.43.4               SummarizedExperiment_1.33.3 Biobase_2.63.0             </span></span>
<span><span class="co">#&gt; [13] MatrixGenerics_1.15.0       matrixStats_1.2.0           GenomicRanges_1.55.3       </span></span>
<span><span class="co">#&gt; [16] GenomeInfoDb_1.39.9         IRanges_2.37.1              S4Vectors_0.41.4           </span></span>
<span><span class="co">#&gt; [19] BiocGenerics_0.49.1         macrophage_1.19.0           BiocStyle_2.31.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] fs_1.6.3                bitops_1.0-7            fontawesome_0.5.2      </span></span>
<span><span class="co">#&gt;   [4] httr_1.4.7              webshot_0.5.5           RColorBrewer_1.1-3     </span></span>
<span><span class="co">#&gt;   [7] doParallel_1.0.17       Rgraphviz_2.47.0        dynamicTreeCut_1.63-1  </span></span>
<span><span class="co">#&gt;  [10] tippy_0.1.0             tools_4.4.0             utf8_1.2.4             </span></span>
<span><span class="co">#&gt;  [13] R6_2.5.1                DT_0.32                 lazyeval_0.2.2         </span></span>
<span><span class="co">#&gt;  [16] GetoptLong_1.0.5        withr_3.0.0             prettyunits_1.2.0      </span></span>
<span><span class="co">#&gt;  [19] gridExtra_2.3           cli_3.6.2               textshaping_0.3.7      </span></span>
<span><span class="co">#&gt;  [22] TSP_1.2-4               sass_0.4.8              bs4Dash_2.3.3          </span></span>
<span><span class="co">#&gt;  [25] genefilter_1.85.1       ggridges_0.5.6          pkgdown_2.0.7          </span></span>
<span><span class="co">#&gt;  [28] systemfonts_1.0.6       AnnotationForge_1.45.0  limma_3.59.5           </span></span>
<span><span class="co">#&gt;  [31] rstudioapi_0.15.0       RSQLite_2.3.5           visNetwork_2.1.2       </span></span>
<span><span class="co">#&gt;  [34] generics_0.1.3          GOstats_2.69.0          shape_1.4.6.1          </span></span>
<span><span class="co">#&gt;  [37] crosstalk_1.2.1         dplyr_1.1.4             dendextend_1.17.1      </span></span>
<span><span class="co">#&gt;  [40] Matrix_1.6-5            fansi_1.0.6             abind_1.4-5            </span></span>
<span><span class="co">#&gt;  [43] lifecycle_1.0.4         yaml_2.3.8              SparseArray_1.3.4      </span></span>
<span><span class="co">#&gt;  [46] BiocFileCache_2.11.1    grid_4.4.0              blob_1.2.4             </span></span>
<span><span class="co">#&gt;  [49] promises_1.2.1          crayon_1.5.2            shinydashboard_0.7.2   </span></span>
<span><span class="co">#&gt;  [52] miniUI_0.1.1.1          lattice_0.22-5          ComplexUpset_1.3.3     </span></span>
<span><span class="co">#&gt;  [55] annotate_1.81.2         KEGGREST_1.43.0         pillar_1.9.0           </span></span>
<span><span class="co">#&gt;  [58] knitr_1.45              ComplexHeatmap_2.19.0   rjson_0.2.21           </span></span>
<span><span class="co">#&gt;  [61] codetools_0.2-19        glue_1.7.0              data.table_1.15.2      </span></span>
<span><span class="co">#&gt;  [64] vctrs_0.6.5             png_0.1-8               gtable_0.3.4           </span></span>
<span><span class="co">#&gt;  [67] assertthat_0.2.1        cachem_1.0.8            xfun_0.42              </span></span>
<span><span class="co">#&gt;  [70] S4Arrays_1.3.6          mime_0.12               survival_3.5-8         </span></span>
<span><span class="co">#&gt;  [73] pheatmap_1.0.12         seriation_1.5.4         iterators_1.0.14       </span></span>
<span><span class="co">#&gt;  [76] statmod_1.5.0           ellipsis_0.3.2          Category_2.69.0        </span></span>
<span><span class="co">#&gt;  [79] bit64_4.0.5             threejs_0.3.3           progress_1.2.3         </span></span>
<span><span class="co">#&gt;  [82] filelock_1.0.3          bslib_0.6.1             colorspace_2.1-0       </span></span>
<span><span class="co">#&gt;  [85] DBI_1.2.2               tidyselect_1.2.1        bit_4.0.5              </span></span>
<span><span class="co">#&gt;  [88] compiler_4.4.0          curl_5.2.1              httr2_1.0.0            </span></span>
<span><span class="co">#&gt;  [91] expm_0.999-9            xml2_1.3.6              desc_1.4.3             </span></span>
<span><span class="co">#&gt;  [94] DelayedArray_0.29.9     plotly_4.10.4           colourpicker_1.3.0     </span></span>
<span><span class="co">#&gt;  [97] bookdown_0.38           scales_1.3.0            RBGL_1.79.0            </span></span>
<span><span class="co">#&gt; [100] NMF_0.27                rappdirs_0.3.3          stringr_1.5.1          </span></span>
<span><span class="co">#&gt; [103] digest_0.6.35           shinyBS_0.61.1          rmarkdown_2.26         </span></span>
<span><span class="co">#&gt; [106] ca_0.71.1               XVector_0.43.1          htmltools_0.5.7        </span></span>
<span><span class="co">#&gt; [109] pkgconfig_2.0.3         base64enc_0.1-3         highr_0.10             </span></span>
<span><span class="co">#&gt; [112] dbplyr_2.4.0            fastmap_1.1.1           rlang_1.1.3            </span></span>
<span><span class="co">#&gt; [115] GlobalOptions_0.1.2     htmlwidgets_1.6.4       shiny_1.8.0            </span></span>
<span><span class="co">#&gt; [118] farver_2.1.1            jquerylib_0.1.4         jsonlite_1.8.8         </span></span>
<span><span class="co">#&gt; [121] BiocParallel_1.37.1     RCurl_1.98-1.14         magrittr_2.0.3         </span></span>
<span><span class="co">#&gt; [124] GenomeInfoDbData_1.2.11 patchwork_1.2.0         munsell_0.5.0          </span></span>
<span><span class="co">#&gt; [127] Rcpp_1.0.12             shinycssloaders_1.0.0   viridis_0.6.5          </span></span>
<span><span class="co">#&gt; [130] stringi_1.8.3           rintrojs_0.3.4          zlibbioc_1.49.3        </span></span>
<span><span class="co">#&gt; [133] MASS_7.3-60.2           plyr_1.8.9              parallel_4.4.0         </span></span>
<span><span class="co">#&gt; [136] ggrepel_0.9.5           bigmemory.sri_0.1.8     Biostrings_2.71.4      </span></span>
<span><span class="co">#&gt; [139] splines_4.4.0           hms_1.1.3               circlize_0.4.16        </span></span>
<span><span class="co">#&gt; [142] locfit_1.5-9.9          igraph_2.0.3            uuid_1.2-0             </span></span>
<span><span class="co">#&gt; [145] rngtools_1.5.2          reshape2_1.4.4          biomaRt_2.59.1         </span></span>
<span><span class="co">#&gt; [148] XML_3.99-0.16.1         evaluate_0.23           BiocManager_1.30.22    </span></span>
<span><span class="co">#&gt; [151] foreach_1.5.2           tweenr_2.0.3            httpuv_1.6.14          </span></span>
<span><span class="co">#&gt; [154] backbone_2.1.3          tidyr_1.3.1             purrr_1.0.2            </span></span>
<span><span class="co">#&gt; [157] polyclip_1.10-6         heatmaply_1.5.0         clue_0.3-65            </span></span>
<span><span class="co">#&gt; [160] ggplot2_3.5.0           gridBase_0.4-7          ggforce_0.4.2          </span></span>
<span><span class="co">#&gt; [163] xtable_1.8-4            later_1.3.2             viridisLite_0.4.2      </span></span>
<span><span class="co">#&gt; [166] ragg_1.3.0              tibble_3.2.1            memoise_2.0.1          </span></span>
<span><span class="co">#&gt; [169] registry_0.5-1          cluster_2.1.6           shinyWidgets_0.8.2     </span></span>
<span><span class="co">#&gt; [172] shinyAce_0.4.2          GSEABase_1.65.1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Alasoo2018" class="csl-entry">
Alasoo, Kaur, Julia Rodrigues, Subhankar Mukhopadhyay, Andrew J Knights,
Alice L Mann, Kousik Kundu, Christine Hale, Gordon Dougan, and Daniel J
Gaffney. 2018. <span>“<span class="nocase">Shared genetic effects on
chromatin and gene expression indicate a role for enhancer priming in
immune response</span>.”</span> <em>Nature Genetics</em> 50 (3): 424–31.
<a href="https://doi.org/10.1038/s41588-018-0046-7" class="external-link">https://doi.org/10.1038/s41588-018-0046-7</a>.
</div>
<div id="ref-Gillespie2022" class="csl-entry">
Gillespie, Marc, Bijay Jassal, Ralf Stephan, Marija Milacic, Karen
Rothfels, Andrea Senff-Ribeiro, Johannes Griss, et al. 2022.
<span>“<span class="nocase">The reactome pathway knowledgebase
2022</span>.”</span> <em>Nucleic Acids Research</em> 50 (D1): D687–92.
<a href="https://doi.org/10.1093/nar/gkab1028" class="external-link">https://doi.org/10.1093/nar/gkab1028</a>.
</div>
<div id="ref-Kanehisa2023" class="csl-entry">
Kanehisa, Minoru, Miho Furumichi, Yoko Sato, Masayuki Kawashima, and
Mari Ishiguro-Watanabe. 2023. <span>“<span class="nocase">KEGG for
taxonomy-based analysis of pathways and genomes</span>.”</span>
<em>Nucleic Acids Research</em> 51 (D1): D587–92. <a href="https://doi.org/10.1093/nar/gkac963" class="external-link">https://doi.org/10.1093/nar/gkac963</a>.
</div>
<div id="ref-Love2014" class="csl-entry">
Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. <span>“<span class="nocase">Moderated estimation of fold change and dispersion for
RNA-seq data with DESeq2</span>.”</span> <em>Genome Biology</em> 15
(12): 550. <a href="https://doi.org/10.1186/s13059-014-0550-8" class="external-link">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-Yoon2019" class="csl-entry">
Yoon, Sora, Jinhwan Kim, Seon-kyu Kim, Bukyung Baik, Sang-mun Chi, and
Seon-young Kim. 2019. <span>“<span class="nocase">GScluster :
network-weighted gene-set clustering analysis</span>,”</span> 1–14.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Annekathrin Nedwed, <a href="https://federicomarini.github.io" class="external-link">Federico Marini</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
